src/GraphQL.MemoryCache/MemoryDocumentCache.cs

```diff
@@ -0,0 +1,58 @@
+using System;
+using GraphQL.Language.AST;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+
+namespace GraphQL.Caching
+{
+    /// <summary>
+    /// A basic implementation of a document cache, limited by a configured amount of memory.
+    /// </summary>
+    public class MemoryDocumentCache : IDocumentCache, IDisposable
+    {
+        private readonly MemoryCache _memoryCache;
+        private readonly TimeSpan _slidingExpiration;
+
+        /// <summary>
+        /// Initializes a new instance with the specified parameters.
+        /// </summary>
+        /// <param name="maxTotalQueryLength">The total length of all queries cached in this instance. Assume maximum memory used is about 10x this value. Will not cache queries larger than 1/3 of this value. During cache compression, reduces cache size by 1/3 of this value.</param>
+        /// <param name="slidingExpiration">The maximum lifetime of queries cached within this instance.</param>
+        public MemoryDocumentCache(int maxTotalQueryLength, TimeSpan slidingExpiration)
```

Comment:
```suggestion
        public MemoryDocumentCache(IMemoryCache cache, MemoryDocumentCacheOptions options)
```

`IMemoryCache` is external (not owned) dependency. In fact it can be even distributed cache.

So `Dispose` method is not needed.

Reply:
A distributed cache cannot contain managed instances of classes, so deserialization is necessary.  At that point we might as well deserialize the query and save the slow network calls!  It certainly would be faster.  Besides, a distributed cache would not implement `IMemoryCache`; it would implement `IDistributedCache`.

https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.distributed.idistributedcache?view=dotnet-plat-ext-5.0

Reply:
I said wrong, forget about `IDistributedCache`. My idea is about taking `IMemoryCache` as dependency so allowing _DI-friendly_ style.

Reply:
The challenging thing about sharing a `IMemoryCache` instance, besides the fact that there can be key collisions, is that the size parameter of entries relates only to how the cache was initialized.  For instance, here the number of bytes allocated is about 10x larger than the value provided to `Size`, due to the amount of memory consumed by the AST copy of the query.  Other users of a cache may use a `Size` value as a simple object counter (all entries have a `Size` of 1).  And other users may have a `Size` value that relates closely to the amount of memory required.  So typically `IMemoryCache` instances are not shared unless the `Size` parameter is not used (which is not recommended due to easy abuse by DOS attacks).

Finally, dictionary lookups should be faster with a private memory cache versus a shared one.  I really don't see a reason to allow an external `IMemoryCache`.  I can add a protected constructor in case someone wants to customize behavior.

Reply:
Well...OK, then let's do that:
```c#
 public class MemoryDocumentCacheOptions : MemoryCacheOptions, IOptions<MemoryDocumentCacheOptions>
        {
            public TimeSpan SlidingExpiration { get; set; }

            MemoryDocumentCacheOptions IOptions<MemoryDocumentCacheOptions>.Value => this;
        }

        public MemoryDocumentCache(MemoryDocumentCacheOptions options) { }

        public MemoryDocumentCache(IOptions<MemoryDocumentCacheOptions> options) { }
```

to be DI-friendly and resistant to adding new options.

Reply:
I'm in the middle of a sample; take a look first; hold on a sec, ok?

Reply:
👍

Reply:
Ok I just pushed some protected members.  Take a look first ok?

Reply:
> I said wrong, forget about `IDistributedCache`. My idea is about taking `IMemoryCache` as dependency so allowing _DI-friendly_ style.

So I think you're suggesting we move the required options to a new options class so it's DI friendly.  That makes sense.

Reply:
Yes. I would like to keep `GetMemoryCacheEntryOptions` method but remove all protected ctors in favor of ` public MemoryDocumentCache(IOptions<MemoryDocumentCacheOptions> options) { }` since this one ctor covers all cases.

Reply:
Ok, options class done.  We can change the additional constructors to `private`, or leave them `public`.  There's also the `protected` members if someone wants to customize behavior further.

Reply:
> Yes. I would like to keep `GetMemoryCacheEntryOptions` method but remove all protected ctors in favor of ` public MemoryDocumentCache(IOptions<MemoryDocumentCacheOptions> options) { }` since this one ctor covers all cases.

I will remove the other `public` constructors then, leaving the `protected` constructors, as they are necessary to override behavior.

Reply:
Done.  I could remove the `protected` members also, but they shouldn't hurt anything.